cli/tfsource/types.go

```diff
@@ -0,0 +1,214 @@
+package tfsource
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/hashicorp/go-getter"
+	urlhelper "github.com/hashicorp/go-getter/helper/url"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gruntwork-io/terragrunt/errors"
+	"github.com/gruntwork-io/terragrunt/util"
+)
+
+var forcedRegexp = regexp.MustCompile(`^([A-Za-z0-9]+)::(.+)$`)
+
+// This struct represents information about Terraform source code that needs to be downloaded
+type TerraformSource struct {
+	// A canonical version of RawSource, in URL format
+	CanonicalSourceURL *url.URL
+
+	// The folder where we should download the source to
+	DownloadDir string
+
+	// The folder in DownloadDir that should be used as the working directory for Terraform
+	WorkingDir string
+
+	// The path to a file in DownloadDir that stores the version number of the code
+	VersionFile string
+}
+
+func (src *TerraformSource) String() string {
+	return fmt.Sprintf("TerraformSource{CanonicalSourceURL = %v, DownloadDir = %v, WorkingDir = %v, VersionFile = %v}", src.CanonicalSourceURL, src.DownloadDir, src.WorkingDir, src.VersionFile)
+}
+
+// Encode a version number for the given source. When calculating a version number, we simply take the query
```

Comment:
Per your inspiration Yori 😄 

```suggestion
// Encode a version number for the given source. When calculating a version number, we take the query
```

Reply:
Ah good catch! I was blindly copy pasting so didn't really comb through these. 👍